dns/roundtrip.go

```diff
@@ -0,0 +1,331 @@
+	if len(buf) > maxMsgSize {
+		return nil, fmt.Errorf("message too large: %v bytes", len(buf))
+	}
+	binary.BigEndian.PutUint16(buf[:2], uint16(len(buf)-2))
+	// TODO: Consider writer.ReadFrom(net.Buffers) in case the writer is a TCPConn.
+	if _, err := conn.Write(buf); err != nil {
+		return nil, fmt.Errorf("failed to write message: %w", err)
+	}
+	var msgLen uint16
+	if err := binary.Read(conn, binary.BigEndian, &msgLen); err != nil {
+		return nil, fmt.Errorf("failed to read message length: %v", err)
+	}
+	if int(msgLen) <= cap(buf) {
+		buf = buf[:msgLen]
+	} else {
+		buf = make([]byte, msgLen)
+	}
+	if _, err = io.ReadFull(conn, buf); err != nil {
+		return nil, fmt.Errorf("failed to read message: %w", err)
+	}
+	var msg dnsmessage.Message
+	if err = msg.Unpack(buf); err != nil {
+		return nil, fmt.Errorf("failed to unpack DNS response: %w", err)
+	}
+	if err := checkResponse(id, q, msg.Header, msg.Questions); err != nil {
+		return nil, fmt.Errorf("invalid response: %w", err)
+	}
+	return &msg, nil
+}
+
+// Implements a DNS exchange over a datagram protocol.
+func dnsPacketRoundtrip(conn io.ReadWriter, q dnsmessage.Question) (*dnsmessage.Message, error) {
+	id := uint16(rand.Uint32())
+	buf, err := appendRequest(id, q, make([]byte, 0, 512))
+	if err != nil {
+		return nil, err
+	}
+	if len(buf) > maxMsgSize {
+		return nil, fmt.Errorf("message too large: %v bytes", len(buf))
+	}
+	if _, err := conn.Write(buf); err != nil {
+		return nil, fmt.Errorf("failed to write message: %w", err)
+	}
+	if cap(buf) >= maxDNSPacketSize {
+		buf = buf[:maxDNSPacketSize]
+	} else {
+		buf = make([]byte, maxDNSPacketSize)
+	}
+	for {
+		n, err := conn.Read(buf)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read message: %w", err)
+		}
+		buf = buf[:n]
+		var msg dnsmessage.Message
+		if err = msg.Unpack(buf); err != nil {
+			return nil, fmt.Errorf("failed to unpack DNS response: %w", err)
+		}
+		if err := checkResponse(id, q, msg.Header, msg.Questions); err != nil {
+			continue
+		}
+		return &msg, nil
+	}
+}
+
+// Maximum DNS packet size.
+// Value taken from https://dnsflagday.net/2020/.
+const maxDNSPacketSize = 1232
+
+// Creates a DNS request using the id and question and appends the bytes to buf.
+func appendRequest(id uint16, q dnsmessage.Question, buf []byte) ([]byte, error) {
+	b := dnsmessage.NewBuilder(buf, dnsmessage.Header{ID: id, RecursionDesired: true})
+	if err := b.StartQuestions(); err != nil {
+		return nil, err
+	}
+	if err := b.Question(q); err != nil {
+		return nil, err
+	}
+
+	// Accept packets up to maxDNSPacketSize.  RFC 6891.
+	if err := b.StartAdditionals(); err != nil {
+		return nil, err
+	}
+	var rh dnsmessage.ResourceHeader
+	if err := rh.SetEDNS0(maxDNSPacketSize, dnsmessage.RCodeSuccess, false); err != nil {
+		return nil, err
+	}
+	if err := b.OPTResource(rh, dnsmessage.OPTResource{}); err != nil {
+		return nil, err
+	}
+
+	buf, err := b.Finish()
+	if err != nil {
+		return nil, err
+	}
+	return buf, nil
```

Comment:
```suggestion
	return b.Finish()
```

Reply:
I added an error msg now, so I need the if.